pkg/controller/direct/secretmanager/secret_controller.go

```diff
@@ -0,0 +1,367 @@
+	annotations := MergeMap(secret.GetAnnotations(), secret.Spec.Annotations)
+	common.RemoveByPrefixes(annotations, "cnrm.cloud.google.com", "alpha.cnrm.cloud.google.com")
+	return annotations
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating Secret", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := SecretManagerSecretSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	resource.Annotations = ComputeAnnotations(desired)
+	resource.Labels = common.ComputeGCPLabels(desired.GetLabels())
+
+	// Add metadata
+	req := &secretmanagerpb.CreateSecretRequest{
+		Parent:   a.id.Parent.String(),
+		SecretId: a.id.Secret,
+		Secret:   resource,
+	}
+	created, err := a.gcpClient.CreateSecret(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Secret %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully created Secret", "name", a.id.FullyQualifiedName())
+
+	status := &krm.SecretManagerSecretStatus{}
+	status.ObservedState = SecretManagerSecretStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = a.id.AsExternalRef()
+	status.Name = created.Name
+	return setStatus(u, status)
+}
+
+func topicsEqual(desired []*krm.TopicRef, actual []*secretmanagerpb.Topic) bool {
+
+	externalsDesired := sets.Set[string]{}
+	externalsActual := sets.Set[string]{}
+	for _, topicRef := range desired {
+		externalsDesired.Insert(topicRef.PubSubTopicRef.External)
+	}
+	for _, topic := range actual {
+		externalsActual.Insert(topic.GetName())
+	}
+	return reflect.DeepEqual(sets.List(externalsDesired), sets.List(externalsActual))
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating Secret", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := SecretManagerSecretSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// the GCP service use *name* to identify the resource.
+	resource.Name = a.id.FullyQualifiedName()
+
+	resource.Annotations = ComputeAnnotations(desired)
+	resource.Labels = common.ComputeGCPLabels(desired.GetLabels())
+
+	updateMask := &fieldmaskpb.FieldMask{}
+
+	if !reflect.DeepEqual(resource.Annotations, a.actual.Annotations) {
+		updateMask.Paths = append(updateMask.Paths, "annotations")
+	}
+	if !topicsEqual(desired.Spec.TopicRefs, a.actual.GetTopics()) {
+		updateMask.Paths = append(updateMask.Paths, "topics")
+	}
+	if !common.DeepEqual_StringAndTimestampPb(*desired.Spec.ExpireTime, a.actual.GetExpireTime()) {
```

Comment:
Nit: It's probably easier to compare resource vs a.actual

Reply:
Yes! exactly what I plan to do next!